internal/infrastructure/kubernetes/serviceaccount.go

```diff
@@ -17,78 +15,43 @@ const (
 	envoyServiceAccountName = "envoy"
 )
 
-// createServiceAccountIfNeeded creates a serviceaccount, if it doesn't exist
-// in the kube api server.
-func (i *Infra) createServiceAccountIfNeeded(ctx context.Context, infra *ir.Infra) error {
-	if infra == nil {
-		return errors.New("infra ir is nil")
-	}
-
-	if infra.Proxy == nil {
-		return errors.New("proxy infra ir is nil")
-	}
-
-	current, err := i.getServiceAccount(ctx)
-	if err != nil {
-		if kerrors.IsNotFound(err) {
-			sa, err := i.createServiceAccount(ctx)
-			if err != nil {
-				return err
-			}
-			if err := i.addResource(sa); err != nil {
-				return err
-			}
-			return nil
-		}
-		return err
-	}
-
-	if err := i.addResource(current); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-// getServiceAccount gets the ServiceAccount from the kube api server.
-func (i *Infra) getServiceAccount(ctx context.Context) (*corev1.ServiceAccount, error) {
-	key := types.NamespacedName{
-		Namespace: i.Namespace,
-		Name:      envoyServiceAccountName,
-	}
-	sa := new(corev1.ServiceAccount)
-	if err := i.Client.Get(ctx, key, sa); err != nil {
-		return nil, fmt.Errorf("failed to get serviceaccount %s/%s: %w",
-			i.Namespace, envoyServiceAccountName, err)
-	}
-
-	return sa, nil
-}
-
 // expectedServiceAccount returns the expected proxy serviceAccount.
 func (i *Infra) expectedServiceAccount() *corev1.ServiceAccount {
 	return &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ServiceAccount",
+			APIVersion: "v1",
+		},
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: i.Namespace,
 			Name:      envoyServiceAccountName,
 		},
 	}
 }
 
-// createServiceAccount creates the Envoy ServiceAccount in the kube api server,
-// if it doesn't exist.
-func (i *Infra) createServiceAccount(ctx context.Context) (*corev1.ServiceAccount, error) {
-	expected := i.expectedServiceAccount()
-	err := i.Client.Create(ctx, expected)
+// createOrUpdateServiceAccount creates the Envoy ServiceAccount in the kube api server,
+// if it doesn't exist and updates it if it does.
+func (i *Infra) createOrUpdateServiceAccount(ctx context.Context, _ *ir.Infra) error {
+	svcAccount := i.expectedServiceAccount()
+	err := i.Client.Create(ctx, svcAccount)
 	if err != nil {
 		if kerrors.IsAlreadyExists(err) {
-			return expected, nil
+			// Update service account if it does not exist
+			if err := i.Client.Update(ctx, svcAccount); err != nil {
+				return fmt.Errorf("failed to update serviceaccount %s/%s: %w",
+					svcAccount.Namespace, svcAccount.Name, err)
+			}
```

Comment:
IMO we should compare expected with actual, e.g. what exists in the controller-runtime cache, and only update if managed fields, e.g. `spec`, change. I'm fine proceeding with the approach in this PR for v0.2.0 as long as a tracking issue is created for doing the comparison.